udata/tests/test_uris.py

```diff
@@ -0,0 +1,294 @@
+]
+
+PUBLIC_HOSTS_IDN = [
+    'http://例子.中国',
+    'http://somewhere.укр',
+]
+
+WITH_CREDENTIALS = [
+    'http://userid:password@example.com:8080',
+    'http://userid:password@example.com:8080/',
+    'http://userid@example.com',
+    'http://userid@example.com/',
+    'http://userid@example.com:8080',
+    'http://userid@example.com:8080/',
+    'http://userid:password@example.com',
+    'http://userid:password@example.com/',
+]
+
+PUBLIC_IPS = [
+    'http://142.42.1.1/',
+    'http://142.42.1.1:8080',
+    'http://142.42.1.1:8080/',
+    'http://223.255.255.254',
+    'http://[2a00:1450:4007:80e::2004]',
+    'http://[2a00:1450:4007:80e::2004]:8080',
+    'http://[2a00:1450:4007:80e::2004]:8080/',
+]
+
+PUBLIC = PUBLIC_HOSTS + PUBLIC_HOSTS_IDN + PUBLIC_IPS + WITH_CREDENTIALS
+
+PRIVATE_IPS = [
+    'http://10.1.1.1',
+    'http://10.1.1.1:8080',
+    'http://10.1.1.1:8080/index.html',
+    'http://10.1.1.254',
+    'http://10.1.1.254:8080',
+    'http://10.1.1.254:8080/index.html',
+    'http://[fc00::1]',
+    'http://[fc00::1]:8080',
+    'http://[fc00::1]:8080/index.html',
+]
+
+PRIVATE = PRIVATE_IPS
+
+LOCAL_HOSTS = [
+    'http://localhost',
+    'http://localhost:8080',
+    'http://localhost:8080/index.html',
+    'http://localhost.localdomain',
+    'http://localhost.localdomain:8080',
+    'http://localhost.localdomain:8080/index.html',
+]
+
+LOCAL_IPS = [
+    'http://127.0.0.1',
+    'http://127.0.0.1:8080',
+    'http://127.0.0.1:8080/index.html',
+    'http://127.0.1.1',
+    'http://127.0.1.1:8080',
+    'http://127.0.1.1:8080/index.html',
+    'http://[::1]',
+    'http://[::1]:8080',
+    'http://[::1]:8080/index.html',
+]
+
+LOCAL = LOCAL_HOSTS + LOCAL_IPS
+
+MULTICAST = [
+    'http://224.1.1.1',
+    'http://224.1.1.1:8080',
+    'http://224.1.1.1:8080/index.html',
+    'http://[ff00::1]',
+    'http://[ff00::1]:8080',
+    'http://[ff00::1]:8080/index.html',
+]
+
+INVALID = [
+    'http://',
+    'h/://.',
+    'http://..',
+    'http://../',
+    'http://?',
+    'http://??',
+    'http://??/',
+    'http://#',
+    'http://##',
+    'http://##/',
+    'http://foo.bar?q=Spaces should be encoded',
+    '//',
+    '//a',
+    '///a',
+    '///',
+    'http:///a',
+    'foo.com',
+    'rdar://1234',
+    'h://test',
+    'http:// shouldfail.com',
+    ':// should fail',
+    'http://foo.bar/foo(bar)baz quux',
+    'http://-error-.invalid/',
+    'http://_error_.invalid/',
+    'http://a.b--c.de/',
+    'http://-a.b.co',
+    'http://a.b-.co',
+    'http://0.0.0.0',
+    'http://10.1.1.0',
+    'http://10.1.1.255',
+    'http://1.1.1.1.1',
+    'http://123.123.123',
+    'http://3628126748',
+    'http://.www.foo.bar/',
+    'http://www.foo.bar./',
+    'http://.www.foo.bar./',
+    'http://[fffff:1450:4007:80e::2004]',
+    'http://[fffff:1450:4007:80e::2004]:8080',
+    'http://[fffff:1450:4007:80e::2004]:8080/index.html',
+    'http://[::]',
+    'http://[::]:8080',
+    'http://[::]:8080/index.html',
+]
```

Comment:
Ok, I see what you mean when you said bulletproof ;-)

Reply:
:stuck_out_tongue_winking_eye: